### ISSUE TITLE
[REVIEW]: OGRe: An Object-Oriented General Relativity Package for Mathematica

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3416]
**Submitting author:** @bshoshany (<a href="http://orcid.org/0000-0003-2222-127X">Barak Shoshany</a>)
**Repository:** <a href="https://github.com/bshoshany/OGRe" target ="_blank">https://github.com/bshoshany/OGRe</a>
**Version:** v1.6
**Editor:** @VivianePons
**Reviewer:** @kostunin, @amelialdrew
**Archive:** <a href="https://doi.org/10.5281/zenodo.5168868" target="_blank">10.5281/zenodo.5168868</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/068573253d55b405eb934bc915e72134/status.svg)](https://joss.theoj.org/papers/068573253d55b405eb934bc915e72134)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/068573253d55b405eb934bc915e72134"><img src="https://joss.theoj.org/papers/068573253d55b405eb934bc915e72134/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/068573253d55b405eb934bc915e72134/status.svg)](https://joss.theoj.org/papers/068573253d55b405eb934bc915e72134)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/bshoshany/OGRe" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@kostunin & @amelialdrew, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @VivianePons know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/bshoshany/OGRe">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@bshoshany) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/bshoshany/OGRe">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@bshoshany) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @kostunin, @amelialdrew it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.02 s (153.3 files/s, 137489.7 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Mathematica                      1            155            487           1852
Markdown                         2             34              0            162
-------------------------------------------------------------------------------
SUM:                             3            189            487           2014
-------------------------------------------------------------------------------


Statistical information for the repository 'b743ec218688700898811fdc' was
gathered on 2021/06/25.
No commited files with the specified extensions were found.
```

**Comment by whedon** [comment]
PDF failed to compile for issue #3416 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by bshoshany** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03416/joss.03416/10.21105.joss.03416.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03416/joss.03416/10.21105.joss.03416.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
:wave: @amelialdrew, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @kostunin, please update us on how your review is going (this is an automated reminder).

**Comment by kostunin** [comment]
is busy with conference, will come back to review soon, sorry for delay

**Comment by VivianePons** [comment]
Dear @amelialdrew and @kostunin , have you had time to start looking at the paper?

**Comment by kostunin** [comment]
Hi @VivianePons my conference is just finished today, I will have a look next week, sorry for delay again.

**Comment by amelialdrew** [comment]
Hi @VivianePons apologies I have also been very busy but will start next week

**Comment by VivianePons** [comment]
No problem! Just wanted to make sure it was still going on. I'll be taking a summer break myself by the end of the week so no rush

**Comment by amelialdrew** [comment]
Hi @VivianePons, what is the best channel to suggest small changes/additions?

**Comment by VivianePons** [comment]
You can discuss them here or directly on the software repo, that's up to you

**Comment by amelialdrew** [comment]
Hi @bshoshany, this is looking good to me so far, it seems as though it would be useful for students getting to grips with Mathematica/GR and the documentation is very clear. I have a few small questions/comments organised roughly into the checkpoints I haven’t ticked off yet:

- Do you have any examples of where the code has been cited/used for more advanced research projects? 

- Main parallelisation section for me only speeds up by 30-40% and I get an error when running `TSetParallelization[True]`:  

```
Parallelization enabled.
Part: Part 1 of {} does not exist
```

Other performance/optimisation claims throughout e.g. ‘Built with speed and performance in mind using optimized algorithms’ are not backed up with numbers, but this may not be necessary.

- Target audience is in Summary rather than Statement of Need, I expect this is fine.

- I think it would be worth citing the references throughout rather than just the list at the end. I would also suggest including a review of other GR/tensor manipulation codes, e.g. https://link.springer.com/article/10.1007/s41114-018-0015-6. Also, the link for the first review on the paper doesn’t work.


Willing to discuss all comments as I am not yet an experienced reviewer.

**Comment by bshoshany** [comment]
Hi @amelialdrew,

Thanks so much for reviewing my package and for your very helpful comments! Here are my replies:

* Unfortunately I do not have any examples of this code being used in research yet, except for my own research. The package is only a few months old and it will probably take some time before people who are using it produce any papers with it. However, it is currently being used by both myself and 4 of my students, and I hope to publish the first paper citing the package in a couple of months.
* How many cores does your system have? The speedup in the documentation is with 12 cores and 8 kernels. If you have less than that, then the speedup will be lower.
* When `TSetParallelization` is turned on, it tries to start all available kernels, but executing `LaunchKernels[]` would fail if some kernels are already running, so it first determines how many kernels are already running and then starts only those that are not. You may have found a bug. When you execute `$ConfiguredKernels`, what do you see? It should be something like `{<<N local kernels>>}` and `$ConfiguredKernels[[1, 1]]` should return the total number of kernels available. From your error message I suspect `$ConfiguredKernels` does not return a `List` on your system. Which version of Mathematica are you using?
* The speedup from the parallelization is probably the only one that can be directly measured. But when I say I built it with speed and performance in mind, I am also referring to many other things happening behind the scene. For example:
1. The components of the tensor in any particular representation are not calculated until they are needed. Once they are calculated, they are stored permanently within the tensor object so they don't need to be calculated again later.
2. If you're trying to calculate e.g. the Ricci tensor, then OGRe first checks if the Riemann tensor has already been calculated. If so, then it uses the result of that calculation when calculating the Ricci tensor, instead of calculating it from scratch. And similarly for all other curvature tensors.
3. The ability to export and import tensor objects means that in your next Mathematica session you can import the tensors with everything that has been calculated so far (components in different representations, curvature tensors, etc.) instead of having to calculate everything from scratch.
4. And so on.
* If you want, I'm happy to move the target audience from the Summary to the Statement of Need.
* The references I linked to are just general references for differential geometry and general relativity that don't necessarily apply to any specific word or sentence, but to the paper as a whole. However, I can convert them to inline references if you prefer.
* I'm not sure why the ISBN link to the Baez and Munian book isn't working. The ISBN is correct but the link itself was generated by JOSS. I'll replace it with a different ISBN (9789810217297) for which the link does seem to work.
* I'll add a reference to the paper by MacCallum, which indeed seems relevant. 

Let me know which of the above changes you want me to make, and then I'll upload a new version of the paper!

**Comment by VivianePons** [comment]
Dear all, I'm glad to see that the review is going on. I'll be on leave for the next 3 weeks. We'll see where we're at when I come back! Have a nice summer

**Comment by bshoshany** [comment]
Have a great summer @VivianePons!

**Comment by amelialdrew** [comment]
Hi @bshoshany,

Thank you for the quick reply and for the further information regarding parallelisation. My system has 4 cores which partially explains the slower speedup, but I would say not entirely. Running ```$ConfiguredKernels``` I get just ```{}```, so it seems you are right that it is not returning a ```List```. I am using Mathematica 12.3.1. Do you need any further info to help with the bug?

Further to sorting out this parallelisation issue, I would recommend:
- Moving target audience to Statement of Need (for no other reason than it says that specifically on our checklist!)
- Converting some of the more specific references to inline e.g. xACT package, adding the review paper as discussed and fixing the link.

Let me know what you think!

**Comment by amelialdrew** [comment]
@VivianePons Happy holidays :)

**Comment by bshoshany** [comment]
@amelialdrew, are you running the code on the Wolfram Cloud, by any chance? That might explain why `$ConfiguredKernels` returns `{}`. If not, how many kernels does your Mathematica license allow? This can be found under Preferences -> Local Kernels. You will see an option "Limit by license availability (N)" where N is the number of kernels available in your license. And which OS are you using?

In any case, I will fix this bug by simply not launching any kernels in advance if `$ConfiguredKernels` is empty. The kernels will then instead be launched automatically by Mathematica the first time a parallelized task is executed. I will incorporate this fix into the next release, which should be ready within a week, and will also include some cool new features I've been working on, such as calculating the geodesic equation, more compact display of tensor components using `TList` and `TShow`, and more.

I will make the changes to `paper.md` as you requested, and upload it to the `paper` branch when I'm done. I will take that opportunity to also add the additional features from the upcoming v1.6 release to the list of features in the paper. I'll do that once the new release is up.

**Comment by amelialdrew** [comment]
@bshoshany No I am not, although I've just looked under Preferences and it says ```Local Kernels is deprecated```, so I assume this more up to date version of Mathematica has removed them. I am using macOS Big Sur 11.4. The updates sound good!

**Comment by bshoshany** [comment]
Hmm... I also use v12.3.1, and I also see the message "Local Kernels is deprecated", but I also see the number of local kernels in the preferences, and my `$ConfiguredKernels` returns a `List`. I get that on both Windows and Linux (I don't have access to a Mac).

The documentation for `$ConfiguredKernels` says that "As of Version 12.3, `$ConfiguredKernels` has been superseded by `$DefaultKernels`", but since I want the package to be compatible with Mathematica all the way down to version 12.0 (some people do not upgrade because that costs extra), I can't use that.

However, I found a simple solution; `CloseKernels[]; LaunchKernels[]` should launch all the available kernels regardless of how many are already running, so I'll just do that instead of what I do now for `TSetParallelization`.

I can't find any information in the release notes or documentation about the Local Kernels themselves being deprecated. Perhaps there is some new parallelization framework that doesn't use multiple local kernels, but I can find any evidence of that. I might have to contact Wolfram directly to ask what the deprecation means.

I have a few more questions to understand the bug better, if you don't mind:

1. What do yet get when you execute `$MaxLicenseSubprocesses`?
2. Start a fresh Mathematica session (after closing all the currently open Mathematica windows if any) and type `LaunchKernels[]`. The output should be a list of kernels. How many kernels are in the list?

Thanks!

**Comment by amelialdrew** [comment]
Ah I've just realised there's an option for ```Enable Local Kernels```  which by default runs with 4 kernels with 4 CPUs which approximately halves the time vs no parallelisation, so this may fix it. I suppose it would be worth adding a line in the docs to make sure they are enabled.

In case it is still needed, ```$MaxLicenseSubprocesses``` gives me ```16```, and ```LaunchKernels[]``` gives 4 kernels, both after switching on ```Enable Local Kernels```.

I hope this helps!

**Comment by bshoshany** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03416/joss.03416/10.21105.joss.03416.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03416/joss.03416/10.21105.joss.03416.pdf) :page_facing_up: :point_left:

**Comment by bshoshany** [comment]
@amelialdrew: The new release is now up, along with the modifications you requested to the paper in the `paper` branch. I believe I solved the parallelization bug, so please check and let me know if it works now or if you still get an error.

I am happy to answer any additional questions and make additional modifications as needed.

**Comment by amelialdrew** [comment]
Hello @bshoshany, thanks for the new changes. The new version works and for me gives ```16 parallel kernels launched. CPU has 4 cores``` which is more than the default, as I think you intended. It still only speeds up from ~13s to ~8s, but I would guess this is system/version dependent, and as it does speed up I am happy to approve.

**Comment by bshoshany** [comment]
Hi @amelialdrew, I do believe the speedup you get is exactly what I would expect you to get.

The number that counts is the number of **cores**, since that is the number of things the CPU can actually do in parallel. On my computer I have 12 cores but only 8 kernels due to license limitations, which means 4 of my 12 CPU cores are not actually being utilized. With the 8 kernels running in parallel on 8 cores, I get a speedup from ~10s to ~3.5s, which is roughly a 2.9x speedup. (One might expect to get an 8x speedup, but Mathematica's parallelization is very inefficient.)

On your computer your license allows 16 kernels, but that doesn't mean much, because you only have 4 cores, so only 4 of these kernels are actually running in parallel at any given time. Therefore the number of actual parallel kernels you have (4) is **half** the number of actual parallel kernels I have (8). You get a 1.6x speedup, which is roughly **half** the speedup I get, so that's perfectly consistent.

Thank you so much for completing the review! 🙂

**Comment by amelialdrew** [comment]
Great, thank you!

**Comment by VivianePons** [comment]
Hi there,

from what I gather, @amelialdrew it looks like you are done with the review and happy with the paper. Thank yo very much for your time and expertise.

@kostunin do you think you can have a look soon? Thanks

**Comment by VivianePons** [comment]
@whedon check references

**Comment by VivianePons** [comment]
@whedon check references from branch paper

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch paper
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/s41114-018-0015-6 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by VivianePons** [comment]
@bshoshany thanks! Can you make a tagged release of the paper / software (on Zenodo, figshare, or other) and send me the DOI? Be careful to use the same metadata (title, author) as for the paper itself

**Comment by bshoshany** [comment]
@VivianePons I'm not sure I understand, doesn't JOSS generate its own DOI for the paper? The paper proof shows `10.21105/joss.03416` as the DOI.

**Comment by VivianePons** [comment]
JOSS does but I need a tagged release of the software

**Comment by bshoshany** [comment]
I see, no problem. The software already has a DOI from Zenodo: [doi:10.5281/zenodo.4742935](https://doi.org/10.5281/zenodo.4742935)

Is that okay, or do I need to create a separate one with the same title as the paper?

**Comment by VivianePons** [comment]
I see that this one is from August 8, has there been no change since then? 

In any case, I need the same title indeed. I believe you can change it without making a new one

**Comment by bshoshany** [comment]
The latest version of OGRe is indeed v1.6 from 2021-08-07. I did change `paper.md` a few days ago due to @kostunin's comments, but that's not reflected in Zenodo since it only refers to the software and not the paper.

I changed the title now. Please let me know if you also need the paper to be included in the tagged release, or if just the software is enough.

**Comment by VivianePons** [comment]
That's perfect, thanks!

**Comment by VivianePons** [comment]
@whedon set 10.5281/zenodo.5168868 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.5168868" target="_blank">10.5281/zenodo.5168868</a> is the archive.

**Comment by VivianePons** [comment]
@whedon set v1.6 as version

**Comment by whedon** [comment]
OK. v1.6 is the version.

**Comment by VivianePons** [comment]
Hi @kostunin you forgot to check the "Performance" box. The box needs to be checked even if there is no performance claim.

**Comment by kostunin** [comment]
> Hi @kostunin you forgot to check the "Performance" box. The box needs to be checked even if there is no performance claim.

done

**Comment by VivianePons** [comment]
I can now propose the paper for publication! Congratulation @bshoshany and thank you @kostunin and @amelialdrew for the reviews

**Comment by VivianePons** [comment]
@whedon recommend-accept from branch paper

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/s41114-018-0015-6 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2552

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2552, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true from branch paper 
```

**Comment by kyleniemeyer** [comment]
Hi @bshoshany, I'm the AEIC on duty this week, doing some final checks. I just have a few minor edits for your paper before accepting:
- [ ] Could you add a bit more to your affiliation? Usually we see at least department/unit in addition to institution
- [ ] Throughout the article, you used hyphens (`-`) as em-dashes (`—`). Can you replace those with the latter? For example, in the first paragraph, "...general relativity - the field..." should be "...general relativity—the field...".
- [ ] At the end of the statement of need section, if you are including the references in the actual sentence, then I think you should use the `@author:2001` citation format rather than `[@author:2001]`. This would change (e.g.) "differential geometry in (Kobayashi & Nomizu, 1996a)" to "differential geometry in Kobayashi & Nomizu (1996a)". The other option is to remove the "in" before the references and combine, like "differential geometry (Kobayashi & Nomizu, 1996a; Kobayashi & Nomizu, 1996b)" by changing the text to `differential geometry [@Kobayashi1; @Kobayashi2]`.

**Comment by bshoshany** [comment]
Hi @kyleniemeyer, thanks for the comments! I have applied all of your suggestions to the paper. Please let me know if any other changes are needed.

**Comment by bshoshany** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03416/joss.03416/10.21105.joss.03416.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03416/joss.03416/10.21105.joss.03416.pdf) :page_facing_up: :point_left:

**Comment by kyleniemeyer** [comment]
@whedon accept deposit=true from branch paper

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1433061143371730944) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2564
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03416](https://doi.org/10.21105/joss.03416)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by kyleniemeyer** [comment]
Congratulations @bshoshany on your article's publication in JOSS!

Many thanks to @kostunin and @amelialdrew for reviewing this, and @VivianePons for editing.

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03416/status.svg)](https://doi.org/10.21105/joss.03416)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03416">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03416/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03416/status.svg
   :target: https://doi.org/10.21105/joss.03416
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03416/status.svg)](https://doi.org/10.21105/joss.03416)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by bshoshany** [comment]
Awesome! Thank you so much @amelialdrew, @kostunin, @VivianePons, @kyleniemeyer, and last but not least, @whedon :)

**Comment by amelialdrew** [comment]
No problem, congratulations!
